src/Common/src/System/Runtime/CompilerServices/PreserveDependencyAttribute.cs

```diff
@@ -0,0 +1,60 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+// TODO https://github.com/dotnet/corefx/issues/41201: Design and expose this publicly.
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>States a dependency that one member has on another.</summary>
+    /// <remarks>
+    /// This can be used to inform tooling of a dependency that is otherwise not evident purely from
+    /// metadata and IL, for example a member relied on via reflection.
+    /// </remarks>
+    [AttributeUsage(
+        AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Field /* AttributeTargets.Class | AttributeTargets.Struct */, // TODO: https://github.com/mono/linker/issues/797
+        AllowMultiple = true, Inherited = false)]
+    internal sealed class PreserveDependencyAttribute : Attribute
+    {
+        /// <summary>Initializes the attribute.</summary>
+        /// <param name="memberSignature">The signature of the member depended.</param>
+        public PreserveDependencyAttribute(string memberSignature)
+        {
+            MemberSignature = memberSignature;
+        }
+
+        /// <summary>Initializes the attribute.</summary>
+        /// <param name="memberSignature">The signature of the member depended on.</param>
+        /// <param name="typeName">The full name of the type containing <paramref name="memberSignature"/>.</param>
+        public PreserveDependencyAttribute(string memberSignature, string typeName)
+        {
+            MemberSignature = memberSignature;
+            TypeName = typeName;
+        }
+
+        /// <summary>Initializes the attribute.</summary>
+        /// <param name="memberSignature">The signature of the member depended on.</param>
+        /// <param name="typeName">The full name of the type containing <paramref name="memberSignature"/>.</param>
+        /// <param name="assemblyName">The name of the assembly containing <paramref name="typeName"/>.</param>
+        public PreserveDependencyAttribute(string memberSignature, string typeName, string assemblyName)
+        {
+            MemberSignature = memberSignature;
+            TypeName = typeName;
+            AssemblyName = assemblyName;
+        }
+
+        /// <summary>Gets or sets the signature of the member depended on.</summary>
+        public string MemberSignature { get; set; }
```

Comment:
I don't think we need to be creative at this point. The setters are an unnecessary complication at this point.

Reply:
I'm not being creative.  This is just following our guidance for what attributes look like, and my hope is that we make this public, so I'm including what I think it would look like based on what we know today.  Obviously it can be changed in the future.

Reply:
What is the guideline? I see tons of attributes in BCL with getter only.

Reply:
Oh, you were commenting on the setters... I thought you were commenting on the properties in general.  I can remove the setters for the first three.

Reply:
The design guidelines are here: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/attributes . The 3 different constructors seems to violate the "AVOID overloading custom attribute constructors."

Reply:
As far as I can tell, the linker currently depends on these constructors existing. We can revisit as part of deciding what we actually want this to look like.